File: tools/ez_tools/registry_parser.py
import os
import pandas as pd
from utils.discovery import find_artifact_files, load_csv_with_progress
from collector.collector import add_rows
from utils.logger import print_and_log
from utils.summary import track_summary

def process_registry(ez_dir: str, batch_size: int, base_dir: str):
    artifact_name = "Registry"
    print_and_log(f"[{artifact_name}] Scanning for relevant CSVs under: {ez_dir}")

    reg_files = find_artifact_files(ez_dir, base_dir, artifact_name)

    if not reg_files:
        print_and_log(f"[{artifact_name}] No Registry CSVs found.")
        return

    for file_path in reg_files:
        print_and_log(f"[{artifact_name}] Processing: {file_path}")
        total_rows = 0
        try:
            for df in load_csv_with_progress(file_path, batch_size, artifact_name=artifact_name):
                timeline_data = []

                for _, row in df.iterrows():
                    timestamp = row.get("LastWriteTimestamp", "")
                    dt = pd.to_datetime(timestamp, utc=True, errors="coerce")
                    if pd.isnull(dt):
                        continue
                    dt_str = dt.isoformat().replace("+00:00", "Z")

                    timeline_row = {
                        "DateTime": dt_str,
                        "TimestampInfo": "Last Write",
                        "ArtifactName": artifact_name,
                        "Tool": "EZ Tools",
                        "Description": row.get("Category", ""),
                        "DataDetails": row.get("Description", ""),
                        "DataPath": row.get("ValueData", ""),
                        "EvidencePath": os.path.relpath(file_path, base_dir) if base_dir else file_path
                    }

                    timeline_data.append(timeline_row)

                total_rows += len(timeline_data)
                add_rows(timeline_data)
                track_summary("EZ Tools", artifact_name, len(timeline_data))
        except Exception as e:
            print_and_log(f"[{artifact_name}] Failed to parse {file_path}: {e}")
            continue

        print_and_log(f"[✓] Parsed {total_rows} timeline rows from: {file_path}")

File: ForensicTimeliner.Core/Tools/EZTools/RegistryParser.cs
using CsvHelper;
using CsvHelper.Configuration;
using ForensicTimeliner.Models;
using ForensicTimeliner.Interfaces;
using ForensicTimeliner.Models;
using ForensicTimeliner.Utils;
using Spectre.Console;
using System.Globalization;

namespace ForensicTimeliner.Tools.EZTools;

public class RegistryParser : IArtifactParser
{
    public List<TimelineRow> Parse(string inputDir, string baseDir, ArtifactDefinition artifact, ParsedArgs args)
    {
        var rows = new List<TimelineRow>();

        Logger.PrintAndLog($"[>] - [{artifact.Artifact}] Scanning for relevant CSVs under: [{inputDir}]", "SCAN");


        var files = Discovery.FindArtifactFiles(inputDir, baseDir, artifact.Artifact);
        if (!files.Any())
        {
            Logger.PrintAndLog($"[!] - [{artifact.Artifact}] No matching files found in: {inputDir}", "WARN");
            return rows;
        }

        foreach (var file in files)
        {
            int timelineCount = 0;
            Logger.PrintAndLog($"[+] - [{artifact.Artifact}] Processing: {Path.GetRelativePath(baseDir, file)}", "PROCESS");

            try
            {
                using var reader = new StreamReader(file);
                using var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)
                {
                    HeaderValidated = null,
                    MissingFieldFound = null
                });

                var records = csv.GetRecords<dynamic>();
                foreach (var record in records)
                {
                    var dict = (IDictionary<string, object>)record;
                    var parsedDt = dict.GetDateTime("LastWriteTimestamp");
                    if (parsedDt == null) continue;

                    string dtStr = parsedDt.Value.ToString("o").Replace("+00:00", "Z");

                    string valueName = dict.GetString("ValueName");
                    string valueData1 = dict.GetString("ValueData");
                    string valueData2 = dict.GetString("ValueData2");
                    string valueData3 = dict.GetString("ValueData3");

                    string description = dict.GetString("Description");
                    string comment = dict.GetString("Comment");
                    string fullDescription = string.IsNullOrWhiteSpace(comment) ? description : $"{description} - {comment}";

                    string keyPath = dict.GetString("KeyPath");
                    string relativePath = Path.GetRelativePath(baseDir, file);
                    string evidencePath = string.IsNullOrWhiteSpace(keyPath) ? relativePath : $"{relativePath} | {keyPath}";

                    rows.Add(new TimelineRow
                    {
                        DateTime = dtStr,
                        TimestampInfo = "Last Write",
                        ArtifactName = "Registry",
                        Tool = artifact.Tool,
                        Description = fullDescription,
                        DataDetails = fullDescription,
                        DataPath = $"{valueName}\\{valueData1}\\{valueData2}\\{valueData3}",
                        EvidencePath = evidencePath
                    });

                    timelineCount++;
                }

                Logger.PrintAndLog($"[✓] - [{artifact.Artifact}] Parsed {timelineCount} timeline rows from: {Path.GetFileName(file)}", "SUCCESS");
                LoggerSummary.TrackSummary(artifact.Tool, artifact.Artifact, timelineCount);
            }
            catch (Exception ex)
            {
                Logger.PrintAndLog($"[{artifact.Artifact}] Failed to parse {file}: {ex.Message}", "ERROR");
            }
        }

        return rows;
    }
}
